# Python – Project


# Description:
The Epic Games Store is a digital video game storefront for Microsoft Windows and macOS, operated by Epic Games. It launched in December 2018 as both a website and a standalone launcher, of which the latter is required to download and play games. The storefront provides a basic catalog, friends list management, matchmaking, and other features. Epic Games has further planned to expand the feature set of the store front.
Solve the below tasks to help Epic Games to expand their business.

In [306]:
import pandas as pd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# Load the data from a CSV file
df = pd.read_csv('c42fa8f5-f6c2-48e6-8521-f0edffe82b3c-epic-games-store----epic-games.csv')




In [256]:
df

game_link  \
0              HITMAN - Game of the Year Edition₹3,249   
1                                     The Medium₹2,949   
2                                       HITMAN 3₹1,889   
3                                  Into the Breach₹446   
4                                  The Lion's Song₹189   
..                                                 ...   
822                                    Beyond Blue₹469   
823  Batman Arkham Asylum Game of the Year Edition₹521   
824                                           Yaga₹745   
825                                   Satisfactory₹745   
826                                          Minit₹372   

                                                name   price  \
0                  HITMAN - Game of the Year Edition  ₹3,249   
1                                         The Medium  ₹2,949   
2                                           HITMAN 3  ₹1,889   
3                                    Into The Breach    ₹446   
4                                    The Lion's Song    ₹189   
..                                               ...     ...   
822                                      Beyond Blue    ₹469   
823  Batman™: Arkham Asylum Game of the Year Edition    ₹521   
824                                             Yaga    ₹745   
825                                     Satisfactory    ₹745   
826                                            Minit    ₹372   

                     developer                                genres  \
0           IO Interactive A/S  ActionAction-AdventureShooterStealth   
1                 Bloober Team                       HorrorAdventure   
2           IO Interactive A/S                         ActionStealth   
3                 Subset Games          Turn-BasedRogue-LiteStrategy   
4               Mi'pu'mi Games                    NarrationAdventure   
..                         ...                                   ...   
822               E-Line Media                  CasualAdventureIndie   
823         Rocksteady Studios                 ActionFightingStealth   
824                Breadcrumbs                        ActionIndieRPG   
825       Coffee Stain Studios      First PersonSimulationOpen World   
826  JW, Kitty, Jukio, and Dom                  IndiePuzzleAdventure   

                                        features                publisher  \
0                                  Single Player       IO Interactive A/S   
1                                  Single Player        Bloober Team S.A.   
2    AchievementsController SupportSingle Player       IO Interactive A/S   
3                                  Single Player             Subset Games   
4                                  Single Player           Mi'pu'mi Games   
..                                           ...                      ...   
822                                Single Player             E-Line Media   
823                                Single Player             Warner Bros.   
824                                Single Player              Versus Evil   
825                           Co-opSingle Player  Coffee Stain Publishing   
826              Controller SupportSingle Player         Devolver Digital   

    date release            initial release on platform   \
0       09/03/20                 Initial Release09/03/20   
1       01/28/21                          WindowsWindows   
2       08/27/20                 Initial Release01/20/21   
3       02/27/18              WindowsWindowsMac OSMac OS   
4       05/13/21              WindowsWindowsMac OSMac OS   
..           ...                                     ...   
822     06/11/20                          WindowsWindows   
823     03/26/10                          WindowsWindows   
824     11/12/19                          WindowsWindows   
825     03/19/19                          WindowsWindows   
826     04/03/18  WindowsWindowsMac OSMac OSMac OSMac OS   

                                        recommended_os  \
0    OS 64

In [257]:
df.shape



(827, 18)

In [258]:
df.columns

Index(['game_link', 'name', 'price', 'developer', 'genres', 'features',
       'publisher', 'date release', 'initial release on platform ',
       'recommended_os', 'cpu', 'memory', 'gpu', 'available achievements',
       'available xp', 'Critics Recommend', 'Top Critic Average',
       'OpenCritic Rating'],
      dtype='object')

In [259]:


df.dtypes



game_link                        object
name                             object
price                            object
developer                        object
genres                           object
features                         object
publisher                        object
date release                     object
initial release on platform      object
recommended_os                   object
cpu                              object
memory                           object
gpu                              object
available achievements           object
available xp                     object
Critics Recommend                object
Top Critic Average              float64
OpenCritic Rating                object
dtype: object

# 1  Epic Games wants to identify the top 2 open critic rating based on the average price. Clean the Price column by removing the rupee symbol from it and store the values in a new column named as ‘Price_Clean’. Then group the data based on average price and ‘OpenCritic Rating’ to find the top 2 open critic ratings.


In [260]:
# Remove rupees  characters and clean the Price column and renamed as Price clean
df["Price_Clean"] = df["price"].str.replace("₹","").str.replace(",","").str.replace("Free","0").astype(float)


In [262]:


df["date release"] = pd.to_datetime(df["date release"])



In [264]:


df.drop(columns=["game_link","available achievements","available xp"], axis=1, inplace=True)



In [265]:
# Drop null values from the 'OpenCritic Rating' column
df = df.dropna(subset=['OpenCritic Rating'])

In [267]:
# Map string ratings to integer values
rating_mapping = {'Strong': 4, 'Mighty': 3, 'Fair': 2, 'Weak': 1}
df['OpenCritic Rating_Integer'] = df['OpenCritic Rating'].map(rating_mapping)

/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/2683486533.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['OpenCritic Rating_Integer'] = df['OpenCritic Rating'].map(rating_mapping)


In [269]:
# Group the data by 'Price_Clean' and calculate the mean of 'Rating_Integer'
grouped_data = df.groupby('Price_Clean')['OpenCritic Rating_Integer'].mean()

In [270]:
# Find the top 2 OpenCritic ratings
top_2_ratings = grouped_data.nlargest(2)

In [271]:
print(top_2_ratings)

Price_Clean
102.0    4.0
119.0    4.0
Name: OpenCritic Rating_Integer, dtype: float64


# 2 On event of the upcoming festive season, Epic Games is planning to provide the games at a discounted price. However, they are not planning to provide the discount for all the games. They will provide discounts only for the games which are in the top 5 price points. For better understanding refer to the table below.
Game Name
Price
A
100
B
120
C
140
D
150
E
200
F
270
G
115
H
150
I
140

In the above table, the games B,C,D,E,F,H and I are the games which be provided in a discounted rate, because they fall under the top 5 price point. 

In [272]:
# Sort the DataFrame by 'Price' in descending order
sorted_df = df.sort_values(by='Price_Clean', ascending=False)

In [273]:
# Select the top 5 price points
top_5_price_points = sorted_df['Price_Clean'].head(5)

In [274]:
# Filter the games which fall within the top 5 price points
discounted_games = df[df['Price_Clean'].isin(top_5_price_points)]

In [275]:
print(discounted_games)

                                    name   price                   developer  \
57                             Days Gone  ₹3,299                 Bend Studio   
122  Horizon Zero Dawn™ Complete Edition  ₹3,299                   Guerrilla   
364                   OCTOPATH TRAVELER™  ₹4,250  Square Enix, ACQUIRE Corp.   
466                         Watch Dogs 2  ₹3,499            Ubisoft Montreal   
474                     Alien: Isolation  ₹3,325           Creative Assembly   

                                 genres                         features  \
57             ActionSurvivalOpen World                    Single Player   
122  ActionRPGAction-AdventureAdventure                    Single Player   
364                                 RPG  Single PlayerController Support   
466                     ActionAdventure               Co-opSingle Player   
474                              Action                    Single Player   

                 publisher date release initial release on pla

# 3 Your job is to extract the top 5 price points and store them in the form of list and reduce 10% from each of them using a lambda function in Python. Then replace the respective values in the Price column.

	•	The genres column contains the genres that are present in the games. It has all the genres written together without any space or special characters. Whatever is the major genre of the game is given first, followed by the other genres. For better understanding refer to the table below.
Game
Genres
A
ActionComedyAdventure
B
AdventureComedy
C
NarrationShooting

In the above table, the major genres for the Games A, B and C are Action, Adventure and Narration respectively.


In [276]:
# Extract the top 5 price points and reduce 10% from each of them
top_5_prices = df['Price_Clean'].nlargest(5)
reduced_prices = top_5_prices.apply(lambda x: x * 0.9)

In [277]:
# Replace the original price values with the updated values
df.loc[df['Price_Clean'].isin(top_5_prices), 'Price_Clean'] = reduced_prices.values

In [278]:
# Function to extract major genre
def extract_major_genre(genres):
    return genres.split(sep='')[0]

In [279]:


import re

# Function to extract major genre
def extract_major_genre(genres):
    return re.split('(?=[A-Z])', genres)[0]

# Convert non-string values in 'Genres' column to strings
df['genres'] = df['genres'].astype(str)

# Extract major genre for each game
df['Major Genre'] = df['genres'].apply(extract_major_genre)



/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/2125099414.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['genres'] = df['genres'].astype(str)
/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/2125099414.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Major Genre'] = df['genres'].apply(extract_major_genre)


In [280]:
print(df)

                                  name   price                  developer  \
0    HITMAN - Game of the Year Edition  ₹3,249         IO Interactive A/S   
1                           The Medium  ₹2,949               Bloober Team   
2                             HITMAN 3  ₹1,889         IO Interactive A/S   
3                      Into The Breach    ₹446               Subset Games   
4                      The Lion's Song    ₹189             Mi'pu'mi Games   
..                                 ...     ...                        ...   
820                      The Messenger    ₹521           Sabotage Studios   
821                            Celeste    ₹595           Matt Makes Games   
822                        Beyond Blue    ₹469               E-Line Media   
824                               Yaga    ₹745                Breadcrumbs   
826                              Minit    ₹372  JW, Kitty, Jukio, and Dom   

                                   genres  \
0    ActionAction-AdventureSho

# 4 Your job is to extract the major genre for each game and store it in a new column and name the column as “Major Genre”. (Hint: All the genre name starts with uppercase). 

	•	The ‘date release’ column consist of the release date for the games. Using the column identify the average ‘Critics Recommend’ rating of a game releasing on Friday vs average ‘Critics Recommend’ rating of a game releasing on Sunday. 

	•	The ‘features’ column consists of the game type. From the column filter out only ‘Single Player’, ‘Multiplayer’ and ‘Co-op’ and store this in a new dataframe called as ‘features_filtered’. From this new dataframe, find the average ‘memory’ required for each of these three different features.

In [282]:
# Remove '%' symbol from 'Critics Recommend' column
df['Critics Recommend'] = df['Critics Recommend'].str.replace('%', '')

print(df)

                                  name   price                  developer  \
0    HITMAN - Game of the Year Edition  ₹3,249         IO Interactive A/S   
1                           The Medium  ₹2,949               Bloober Team   
2                             HITMAN 3  ₹1,889         IO Interactive A/S   
3                      Into The Breach    ₹446               Subset Games   
4                      The Lion's Song    ₹189             Mi'pu'mi Games   
..                                 ...     ...                        ...   
820                      The Messenger    ₹521           Sabotage Studios   
821                            Celeste    ₹595           Matt Makes Games   
822                        Beyond Blue    ₹469               E-Line Media   
824                               Yaga    ₹745                Breadcrumbs   
826                              Minit    ₹372  JW, Kitty, Jukio, and Dom   

                                   genres  \
0    ActionAction-AdventureSho

/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/154514261.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Critics Recommend'] = df['Critics Recommend'].str.replace('%', '')


In [297]:
# Function to clean and standardize memory values
def clean_memory(memory):
    try:
        # Convert to lowercase and remove non-alphanumeric characters
        memory = memory.lower().replace(' ', '').replace('ram', '')
        # Handle TBD and NAN
        if memory in ['tbd', 'nan']:
            return np.nan
        # Extract numeric value
        numeric_value = float(''.join(filter(str.isdigit, memory)))
        # Convert to GB
        if 'gb' in memory:
            return numeric_value
        elif 'mb' in memory:
            return numeric_value / 1024
        else:
            return np.nan  # Unknown unit
    except:
        return np.nan  # Handle any other errors
# Clean memory column and create memory_cleaned column
df['memory_cleaned'] = df['memory'].apply(clean_memory)

/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/3220688471.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['memory_cleaned'] = df['memory'].apply(clean_memory)


In [298]:

# Extracting major genre and storing it in a new column
df['Major Genre'] = df['genres'].str.split(',').str[0]


/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/4213282568.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Major Genre'] = df['genres'].str.split(',').str[0]


In [299]:
# Converting 'date release' column to datetime format
df['date release'] = pd.to_datetime(df['date release'])

/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/2356899481.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['date release'] = pd.to_datetime(df['date release'])


In [300]:
# Calculating average 'Critics Recommend' rating for games releasing on Friday vs. Sunday
friday_avg_rating = df[df['date release'].dt.dayofweek == 4]['Critics Recommend'].mean()
sunday_avg_rating = df[df['date release'].dt.dayofweek == 6]['Critics Recommend'].mean()

In [301]:
# Filtering out 'Single Player', 'Multiplayer', and 'Co-op' from the 'features' column
features_filtered = df[df['features'].isin(['Single Player', 'Multiplayer', 'Co-op'])]

In [287]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 455 entries, 0 to 826
Data columns (total 18 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   name                          402 non-null    object        
 1   price                         402 non-null    object        
 2   developer                     402 non-null    object        
 3   genres                        455 non-null    object        
 4   features                      352 non-null    object        
 5   publisher                     402 non-null    object        
 6   date release                  402 non-null    datetime64[ns]
 7   initial release on platform   402 non-null    object        
 8   recommended_os                392 non-null    object        
 9   cpu                           391 non-null    object        
 10  memory                        389 non-null    object        
 11  gpu                           37

In [302]:
df.head(50)

name    price  \
3                                  Into The Breach     ₹446   
4                                  The Lion's Song     ₹189   
6                                      Sunless Sea     ₹449   
7                                         Among Us     ₹119   
8                         2064: Read Only Memories     ₹469   
12                                            20XX     ₹477   
14                  Back 4 Blood: Standard Edition   ₹2,999   
16                                  Cyberpunk 2077   ₹2,999   
19                                   House Flipper     ₹569   
21                                    Superliminal     ₹595   
22                           Red Dead Redemption 2   ₹3,199   
23                  Tony Hawk's™ Pro Skater™ 1 + 2   ₹2,526   
24                                Edge Of Eternity  ₹389.95   
25                        Jurassic World Evolution   ₹2,855   
26                           Detroit: Become Human   ₹1,899   
27                                      Townscaper     ₹139   
31                                       The Surge     ₹599   
32                           STAR WARS™: Squadrons   ₹2,499   
33                                       BIOMUTANT   ₹1,889   
34            Conway: Disappearance at Dahlia View     ₹709   
35                                   Port Royale 4   ₹2,299   
36                                  Genshin Impact     Free   
38                        A Plague Tale: Innocence   ₹1,699   
39                           HOT WHEELS UNLEASHED™   ₹1,499   
41           Warhammer Age of Sigmar: Storm Ground     ₹799   
42                   Nioh 2 – The Complete Edition   ₹2,499   
43                       Mafia: Definitive Edition   ₹2,199   
44               Samurai Jack: Battle Through Time     ₹968   
54                                            ELEX   ₹1,099   
55                           Raji: An Ancient Epic     ₹595   
56                                 SAMURAI SHODOWN   ₹1,415   
57                                       Days Gone   ₹3,299   
58                                       MotoGP™21   ₹1,199   
60                         LEGO® Builder's Journey     ₹469   
63                            Alan Wake Remastered     ₹709   
65                                       Tropico 6   ₹2,299   
67                         Hood: Outlaws & Legends     ₹799   
68                                          Vampyr   ₹1,564   
70                                            Myst     ₹709   
71                          New Super Lucky's Tale     ₹709   
73                           Just Cause 4 Reloaded   ₹1,117   
75                                  Darksiders III   ₹2,165   
77                                      Tannenberg     ₹599   
78  The Legend of Heroes: Trails of Cold Steel III   ₹1,299   
79                 Phoenix Point: Year One Edition     ₹709   
80                       Kingdom Come: Deliverance     ₹699   
81                                    Chernobylite     ₹709   
82                                   Void Bastards     ₹639   
85                            JETT : The Far Shore     ₹709   
86                                   In Sound Mind     ₹759   

                       developer                                    genres  \
3                   Subset Games              Turn-BasedRogue-LiteStrategy   
4                 Mi'pu'mi Games                        NarrationAdventure   
6               Failbetter Games                         RPGAdventureIndie   
7                     Innersloth                                    Casual   
8                   MidBoss, LLC                                 Adventure   
12                 Batterystaple                              Rogue-Liteer   
14           Turtle Rock Studios                 ActionFirst PersonShooter   
16                CD PROJEKT RED              ActionRPGOpen WorldAdventure   
19                      Empyrean                     CasualSimulationIndie   
21                 Pillow Castle                    

In [292]:
# Drop NaN values from memory_cleaned column
df.dropna(subset=['memory_cleaned'], inplace=True)

/var/folders/ty/dyrr0z3n5sd444c4606ckkww0000gn/T/ipykernel_1314/1465751088.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.dropna(subset=['memory_cleaned'], inplace=True)


In [303]:
df.columns

Index(['name', 'price', 'developer', 'genres', 'features', 'publisher',
       'date release', 'initial release on platform ', 'recommended_os', 'cpu',
       'memory', 'gpu', 'Critics Recommend', 'Top Critic Average',
       'OpenCritic Rating', 'Price_Clean', 'OpenCritic Rating_Integer',
       'Major Genre', 'memory_cleaned'],
      dtype='object')

In [304]:
# Finding average 'memory' required for each feature
average_memory = features_filtered.groupby('features')['memory_cleaned'].mean()

In [305]:
print("Average memory required for each feature:")
print(average_memory)

Average memory required for each feature:
features
Co-op            3.500000e+00
Multiplayer      1.125000e+01
Single Player    3.385058e+10
Name: memory_cleaned, dtype: float64


# conclusion
# 1
The output indicates that the top 2 OpenCritic ratings are both rated as 4.0. These ratings correspond to the average prices of 102.0 and 119.0, respectively. Therefore, games with these average prices tend to have the highest OpenCritic ratings, as per the provided dataset.

# 2
Games like "Days Gone," "Horizon Zero Dawn™ Complete Edition," "OCTOPATH TRAVELER™," "Watch Dogs 2," and "Alien: Isolation" are eligible for discounts because their prices are among the top 5 highest-priced games. These games will likely be offered at discounted rates during the upcoming festive season by Epic Games.


# 3
The output of the task involves extracting the top 5 price points, reducing each price by 10%, and then replacing the original price values in the DataFrame with the updated values. Additionally, the major genre for each game has been extracted and stored in a new column named 'Major Genre' by splitting the genres string based on capital letters. The warning messages indicate that there might be SettingWithCopyWarning issues, which typically arise when assigning values to a slice of a DataFrame without using .loc. This warning suggests a potential risk of unintentional changes to the original DataFrame. To mitigate this, it's advisable to use .loc for assigning values to avoid any unexpected behavior. Overall, the task has been completed successfully, and the DataFrame now contains the updated price values and a new column indicating the major genre for each game.

# 4
The output displays the average memory required for each feature in a software or game development context. It indicates that "Co-op" requires approximately 3.5 units of memory, "Multiplayer" demands around 11.25 units, and notably, "Single Player" necessitates an unexpectedly large amount of memory, approximately 33.85 billion units. This suggests a potential anomaly or error in the data, as such an immense memory requirement for a single-player feature seems highly unlikely and warrants further investigation or clarification.
